File: core/execution_reality/runtime_conflict_detection_engine.py
from __future__ import annotations

from typing import Any, Dict, List, Tuple


MAX_CONFLICTS = 1000


def detect_runtime_conflicts(
    runtime_ir: Dict[str, Any],
) -> Dict[str, Any]:

    transitions = runtime_ir.get(
        "transitions",
        [],
    )

    seen = set()

    conflicts: List[Tuple[Any, Any]] = []

    for transition in transitions:

        key = (
            transition.get("from"),
            transition.get("to"),
        )

        if key in seen:

            conflicts.append(key)

        seen.add(key)

    return {
        "conflicts": sorted(
            conflicts,
            key=lambda x: (str(x[0]), str(x[1])),
        )[:MAX_CONFLICTS],
    }

File: src/execution_reality/runtimeConflictDetectionEngine.ts
/**
 * Converted from Python: core/execution_reality/runtime_conflict_detection_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let MAX_CONFLICTS: any = 1000;
export function detectRuntimeConflicts(runtime_ir: any): any {
  var transitions: any = py.get(runtime_ir, "transitions", []);
  var seen: Set<any> = new Set();
  var conflicts: any[] = [];
  var transition: any;
  for (transition of py.iter(transitions)) {
    var key: any = [py.get(transition, "from"), py.get(transition, "to")];
    if (py.contains(seen, key)) {
      py.listAppend(conflicts, key);
    }
    py.setAdd(seen, key);
  }
  return {"conflicts": py.slice(py.sorted(conflicts, {key: ((x: any) => [py.toStr(py.at(x, 0)), py.toStr(py.at(x, 1))]) as (item: any) => any}), null, MAX_CONFLICTS)};
}
